components/ConvertSteps/ConvertSteps.js

```diff
@@ -1,98 +1,202 @@
-import React, { useCallback, useState } from 'react'
+import React, { useCallback, useState, useEffect, useReducer } from 'react'
 import PropTypes from 'prop-types'
 import {
   STEPPER_IN_PROGRESS,
   STEPPER_SUCCESS,
   STEPPER_ERROR,
+  STEP_WAITING,
+  STEP_WORKING,
+  STEP_SUCCESS,
+  STEP_ERROR,
 } from './stepper-statuses'
 import Divider from './Divider'
-import ManageStep from './ManageStep'
+import Step from './Step'
 import StepperLayout from './StepperLayout'
 import StepperTitle from './StepperTitle'
+import useStepLayout from './useStepLayout'
+
+function initialStepState(steps) {
+  return steps.map(() => {
+    return {
+      status: STEP_WAITING,
+      hash: null,
+      active: false,
+    }
+  })
+}
+
+function reduceSteps(steps, [action, stepIndex, value]) {
+  if (action === 'setActive') {
+    steps[stepIndex].active = true
+    return [...steps]
+  }
+  if (action === 'setHash') {
+    steps[stepIndex].hash = value
+    return [...steps]
+  }
+  if (action === 'setStatus') {
+    steps[stepIndex].status = value
+    return [...steps]
+  }
+  return steps
+}
 
 function ConvertSteps({ toAnj, fromAmount, toAmount, onReturnHome, steps }) {
   const [stepperStatus, setStepperStatus] = useState(STEPPER_IN_PROGRESS)
   const [stepperStage, setStepperStage] = useState(0)
-  const [retryingStep, setRetryingStep] = useState(null)
+  const [outerBoundsRef, innerBoundsRef, layoutName] = useStepLayout(100)
 
-  const handleRetry = useCallback(() => {
-    setStepperStatus(STEPPER_IN_PROGRESS)
-    setRetryingStep(stepperStage)
-  }, [stepperStage])
+  const [stepState, updateStep] = useReducer(
+    reduceSteps,
+    initialStepState(steps)
+  )
+
+  const attemptStepSigning = useCallback(
```

Comment:
Nice! I really like being able to move this bit of logic all the way up!

The only potential problem here is that any `setState()` after an `await()` may potentially be thrown away, and because this is in a callback, it's difficult for us to properly clean up (rather than say, in an `useEffect()` where we can create a local closure with the `cancel`).

I think it's fine here—you can't unmount this component without closing the app, but in a larger app, this might be something we move up to a context that we can ensure never gets unmounted :).

Reply:
Agreed, It's a pain how verbose `async` management is inside `useEffect`. I'll leave this for now in the interest of time.